pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/formats/protobufnative/PulsarProtobufNativeFormatFactory.java

```diff
@@ -105,17 +105,34 @@ public ChangelogMode getChangelogMode() {
     }
 
     private String extractTopicName(DynamicTableFactory.Context context) {
-
         validateTopic(context.getCatalogTable().getOptions());
-        String topic = context.getCatalogTable().getOptions().get(TOPIC.key());
-        // Maybe catalog mode
-        if (StringUtils.isNullOrWhitespaceOnly(topic)) {
-            final ObjectIdentifier table = context.getObjectIdentifier();
-            topic = TopicName.get(table.getDatabaseName() + "/" + table.getObjectName()).toString();
+
+        String topic;
+        if (isGenericTable(context)) {
+            topic = context.getCatalogTable().getOptions().get(TOPIC.key());
+        } else {
+            String database = context.getObjectIdentifier().getDatabaseName();
+            String objectName = context.getObjectIdentifier().getObjectName();
+            topic = TopicName.get(database + "/" + objectName).toString();
         }
         return topic;
     }
 
+    /**
+     * Check if the table is an generic table. A generic table is defined as created using "CREATE
+     * TABLE". If the table is generic, when getting the table from catalog there will be an GENERIC
+     * option set to True
+     *
+     * @param context
+     * @return true if the topic is created by "CREATE TABLE" , false if the table is directly
+     *     mapped from pulsar topic
+     */
+    private boolean isGenericTable(DynamicTableFactory.Context context) {
+        final String isGeneric = context.getCatalogTable().getOptions().get(GENERIC.key());
```

Comment:
Haven't tested if the GENERIC option will be set.  The GENERIC option is set when call getTable(), but will it be set when creating the table ?

Reply:
The `GENERIC` is set whenever getTable() is called.
And any interaction with the table must call `getTable()` first.